spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/actions/BaseDeleteOrphanFilesSparkAction.java

```diff
@@ -361,4 +457,51 @@ static PathFilter forSpecs(Map<Integer, PartitionSpec> specs) {
       return partitionNames.isEmpty() ? HiddenPathFilter.get() : new PartitionAwareHiddenPathFilter(partitionNames);
     }
   }
+
+  private static MapPartitionsFunction<Row, Row> toFileMetadata(Map<String, String> equalSchemeAndAuthoritiesMap) {
+    return rows -> Iterators.transform(rows, row -> {
+      String filePathAsString = row.getString(0);
+      URI uri = new Path(filePathAsString).toUri();
+      String scheme = equalSchemeAndAuthoritiesMap.getOrDefault(uri.getScheme(), uri.getScheme());
+      String authority = equalSchemeAndAuthoritiesMap.getOrDefault(uri.getAuthority(), uri.getAuthority());
+      return RowFactory.create(scheme, authority, uri.getPath(), filePathAsString);
+    });
+  }
+
+  static class MapOrphanFilesFunction implements MapPartitionsFunction<FileDescriptor, String> {
+
+    private final SetAccumulator<Pair<String, String>> setAcc;
+
+    MapOrphanFilesFunction(SetAccumulator<Pair<String, String>> accumulator) {
+      this.setAcc = accumulator;
+    }
+
+    @Override
+    public Iterator<String> call(Iterator<FileDescriptor> iter) throws Exception {
+      Iterator<String> orphanFilesIterator = Iterators.transform(iter,
+          row -> isOrphan(row) ? row.getActualFilePath() : null);
+      return Iterators.filter(orphanFilesIterator, str -> !Strings.isNullOrEmpty(str));
+    }
+
+    boolean isOrphan(FileDescriptor desc) {
+      if (desc.getValidPath() == null) {
+        return true;
+      }
+      if (!isEqual(desc.getActualScheme(), desc.getValidScheme())) {
+        setAcc.add(Pair.of(desc.getActualScheme(), desc.getValidScheme()));
+      }
+      if (!isEqual(desc.getActualAuthority(), desc.getValidAuthority())) {
+        setAcc.add(Pair.of(desc.getActualAuthority(), desc.getValidAuthority()));
+      }
+      return false;
+    }
+
+    private boolean isEqual(String actual, String valid) {
```

Comment:
It is a little bit hard to wrap the head around this logic. Would something like this be easier to grasp?

```
  private static MapPartitionsFunction<Tuple2<FileMetadata, FileMetadata>, String> findOrphanFiles(
      PrefixMismatchMode mode,
      SetAccumulator<Pair<String, String>> conflicts) {

    return rows -> {
      Iterator<String> transformed = Iterators.transform(rows, row -> {
        FileMetadata actual = row._1;
        FileMetadata valid = row._2;

        if (valid.path == null) {
          return actual.location;
        }

        boolean schemeMatch = Strings.isNullOrEmpty(valid.scheme) || valid.scheme.equalsIgnoreCase(actual.scheme);
        boolean authorityMatch = Strings.isNullOrEmpty(valid.authority) || valid.authority.equalsIgnoreCase(actual.authority);

        switch (mode) {
          ...
        }
      });

      return Iterators.filter(transformed, Objects::nonNull);
    };
  }
```